File: 백준/Gold/15683. 감시/감시.py
import sys
input = sys.stdin.readline
from collections import deque
sys.setrecursionlimit(10**6)

def cal(tlst):
    v = [[0]*M for _ in range(N)]

    # 모든 CCTV에 대해서 처리(좌표, type, rot)
    for i in range(CNT):
        x, y = lst[i]       # 0~N-1, 0~M-1
        rot = tlst[i]       # 0~3
        type = graph[x][y]  # 1~5

        # type에 따른 모든 방향을 뻗어가면서 v[]에 1 표시
        for dr in cctv[type]:
            dr = (dr+rot)%4
            nx, ny = x, y
            while True:
                nx, ny = nx + dx[dr], ny + dy[dr]
                if 0 > nx or nx >= N:
                    break
                if 0 > ny or ny >= M:
                    break
                if graph[nx][ny] == 6:  # 벽이면 멈춤
                    break
                v[nx][ny] = 1

    # 사각지대 개수 카운트
    cnt = 0
    for i in range(N):
        for j in range(M):
            if graph[i][j] == 0 and v[i][j] == 0:
                cnt+=1
    return cnt


def dfs(n, tlst):
    global ans
    if n == CNT:    # 모든 CCTV 방향(0~3) 정하기 완료
        ans = min(ans, cal(tlst))
        return

    dfs(n+1, tlst+[0])  #   0도 회전
    dfs(n+1, tlst+[1])  #  90도 회전
    dfs(n+1, tlst+[2])  # 180도 회전
    dfs(n+1, tlst+[3])  # 270도 회전


N, M = map(int, input().split())

graph = []
# 1번 ~ 5번 CCTV 저장
lst = []
dx = [-1,0,1,0]
dy = [0,1,0,-1]
cctv = [[], [1], [1,3], [0,1], [0,1,3], [0,1,2,3]]
ans = M*N

for _ in range(N):
    graph.append(list(map(int, input().split())))

for i in range(N):
    for j in range(M):
        if 1<=graph[i][j]<=5:
            lst.append([i,j])

CNT = len(lst)
dfs(0,[])
print(ans)

File: 백준/Gold/15683. 감시/감시.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

class Coor {
	int x;
	int y;

	public Coor(int x, int y) {
		super();
		this.x = x;
		this.y = y;
	}

	public int getX() {
		return x;
	}

	public void setX(int x) {
		this.x = x;
	}

	public int getY() {
		return y;
	}

	public void setY(int y) {
		this.y = y;
	}
}

public class Main {
	private static int n, m, cnt, ans;
	private static int[][] graph;
	private static List<Coor> lst;
	private static int[] dx = { -1, 0, 1, 0 };
	private static int[] dy = { 0, 1, 0, -1 };
	private static int[][] cctv = { {}, { 1 }, { 1, 3 }, { 0, 1 }, { 0, 1, 3 }, { 0, 1, 2, 3 } };

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		graph = new int[n][m];
		lst = new ArrayList<>();
		ans = Integer.MAX_VALUE;

		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < m; j++) {
				graph[i][j] = Integer.parseInt(st.nextToken());
				if (1 <= graph[i][j] && graph[i][j] <= 5) {
					lst.add(new Coor(i, j));
				}
			}
		}
		cnt = lst.size();
		dfs(0, new int[cnt]);
		System.out.println(ans);
	}

	private static void dfs(int n, int[] asd) {
		if (n == cnt) {
			ans = Math.min(ans, cal(asd));
			return;
		}
		asd[n] = 0;
		dfs(n + 1, asd);
		asd[n] = 1;
		dfs(n + 1, asd);
		asd[n] = 2;
		dfs(n + 1, asd);
		asd[n] = 3;
		dfs(n + 1, asd);
	}

	private static int cal(int[] tmp) {
		int[][] visited = new int[n][m];

		// 이제 4 방향 회전하면서 체크하면 될 듯!!
		for (int i = 0; i < cnt; i++) {
			int x = lst.get(i).getX();
			int y = lst.get(i).getY();
			int rot = tmp[i];
			int type = graph[x][y];

			for (int dir : cctv[type]) {
				dir = (dir + rot) % 4;
				int nx = x;
				int ny = y;
				while (true) {
					nx = nx + dx[dir];
					ny = ny + dy[dir];
					if (0 > nx || nx >= n)
						break;
					if (0 > ny || ny >= m)
						break;
					if (graph[nx][ny] == 6)
						break;
					visited[nx][ny] = 1;
				}
			}
		}

		int res = 0;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (graph[i][j] == 0 && visited[i][j] == 0)
					res++;
			}
		}
		return res;
	}
}